### ISSUE TITLE
Fix S1905: Rule should not report on Cast<T> over multi-dimensional arrays

### ISSUE THREAD

**Issue by crypto-rsa** [SonarSource/sonar-csharp#825]
### Description

Violation of S1905 (unnecessary cast) is reported when a multidimensional (MD) array of type `T` is cast to `IEnumerable<T>` to allow further LINQ calls. However, since MD arrays do not implement `IEnumerable<T>`, this cast is actually necessary.

### Repro steps

```
var numbers = new int[2, 2];

// this is OK
foreach( int number in numbers )
{
    ...
}

// in order to use the Where extension method, Cast needs to be called first
// this produces S1905 violation
foreach( int number in numbers.Cast<int>().Where( i => i > 0 ) )
{
    ...
}
```

### Expected behavior

S1905 is not reported where a MD array is cast to `IEnumerable<T>` and subsequent calls to extension methods exist.

### Actual behavior

S1905 is reported.

### Related information

* SonarC# Version 3.5.0.1803
* Visual Studio Version 15.3.5

**Comment by Evangelink** [comment]
Hi @crypto-rsa,
Indeed this is a False Positive. Thanks for the feedback!
